docs/csharp/language-reference/operators/null-parameter-check.md

```diff
@@ -0,0 +1,81 @@
+    {
+        throw new ArgumentNullException(nameof(lastName));
+    }
+    // ...
+}
+```
+
+It's intended for library authors to provide runtime checks even when APIs have been annotated for nullable reference types. Some libraries support projects that aren't nullable enabled. Some projects support multiple versions including some that predate nullable reference types. These checks can simplify the necessary validation. You should only add the `!!` operator on those parameters that require a null-check for safety. In the preceding example, the `message` parameter doesn't have a runtime null check.
+
+There are a few rules that govern where you can add the `!!` operator on a parameter declaration:
+
+- The `!!` operator directs the compiler to add runtime behavior. It can't be applied to a declaration without an implementation. You can't add `!!` to the parameter of an abstract member, an interface method without an implementation, or a parameter on a delegate type declaration.
+- The parameter type must be a type that can be compared to `null`. For example, the parameter can't be a non-nullable value type. If the parameter is a type parameter, it can't be constrained to be a non-nullable type (for example the `struct`, `notnull`, and `unmanaged` constraints are disallowed.)
+
+In almost all cases, the code the compiler generates for `!!` on an argument is consistent with the preceding example. However, the null check is inserted before any code you write in the method. The compiler generated null-check in constructors appears before the compiler generated calls to field initializers and before the call to any other constructor or base constructor. You can't catch the <xref:System.ArgumentNullException?displayProperty=nameWithType> thrown by the null check added by the compiler.
+
+## Null parameter checks and nullable types
+
+With null state analysis, any parameter annotated with `!!` has an initial null state of *not null*. The compiler has already added a null check. The compiler generates a warning when you write a method where a parameter is nullable and you've added a null check. Consider the following example:
+
+```csharp
+void Method(string? name!!)
+{
+    // ...
+}
+```
+
+The compiler issues the warning *"Nullable type 'string?' is null-checked and will throw if null."* The signature indicates that the method should accept `null` as the argument for `name`, and yet throws when that argument is supplied.
+
+This feature is intended for library authors whose libraries will be used by code that may not enable nullable reference types. When all components are compiled in a nullable enabled context, the need for extra null parameter checks is reduced, and may not be needed at all. However, when consumers may not have enabled nullable reference types, that code may call methods with null arguments. You can add the runtime checks to improve the resilience of your code. The `!!` syntax makes those extra checks more concise.
```

Comment:
Please drop this paragraph, due the aforementioned reasons.

Especially for libraries one never knows what inputs will arrive at runtime. So no matter if nullable annotations are enabled or not, argument validation has to be done for arguments that disallow null.

Reply:
ad 1)

I think it's a difference whether `NullReferenceException` (NRE) is thrown from anywhere down the stack, or if an `ArgumentNullException` (ANE) is thrown. The latter is part of the "API-contract" of that method. (Therefore ideally it should be documented (xml doc-comments) too).

NRE is kind of unexpected failure, that shouldn't happen. That's IMO why nullability annotations came to C#, so that the compiler can aid in detecting NREs by static analysis.

ANE on the other hand as part of the API-contract can happen. And they are not XOR to NRT, the work together very well.
Example:
```c#
public int Hash_A(string value!!)
{
    return 42;
}

public int Hash-B(string? value)
{
    if (value is null) return 0;
    return 42;
}
```
In `Hash_A` the type is `string` that doesn't allow null. But this "doesn't allow" is only for the static analyis in Roslyn, _not_ for the run-time / execution-time. So argument validation needs to be done -- in the example very handy with the new `!!`-operator.

For `Hash_B` the type is `string?` so for Roslyn's static analysis `null` is allowed, and as part of the API-contract it's also state that `null` can be passed in an no exception should occur in this case. The implementation can handle `null`.

---

ad 2)

> Does NRT do anything useful if I still need to check everything against null?

One only needs to check types that are annotated to not-allow-null (in public APIs).
NRT are still very useful (and I like them very much), as these annotation are part of the signature / API-contract.
So without looking into the docs or xml-doc-comments for a method I can see in the signature if the type allows null or not.
When the type disallows null, and I still pass a null in -- then it's my fault and an ANE should be thrown.
This is my bug, and very easy to track down to fix it.

If there wouldn't be a ANE thrown, rather anywhere and anytime later down the stack a NRE gets throws it's a nightmere to reproduce that bug and cover this by test, in order for a proper fix.
"Fail early" is very helpful.

> If a class library is only accessed from code known to have NRT enabled, is a null check still needed? 

Yes.
NRT is mostyl for static analysis, and being part of the API-contract.
For execution-time the null-check is needed.

I don't think the rules changed / respectively need to be changed.
With NRT we got another useful tool to express the intention: does a type allow to be null / disallows to be null
But this doesn't effect execution-time, there everything remains the same.

> If the only difference is where an exception is thrown, and what type of exception ...

These are big differences. In semantics of the exception, and in the handling of an exception.
The more precise (in terms of type and location) an exception is, the easier it's to deal with it.

Side note: the worst exception is GDI+ generic error
Nobody knows where it comes, and where it goes 😉 

---

So to sum up:
* Runtime exceptions are useful to handle that failure
* NRT are another useful tool to express intent as part of the API-contract
* at execution-time null-checks are still needed (by manual checking or by `!!`)
* NRT and null-checks play together -- they are friends, not opponents